### DISCUSSION TITLE
Deactivating then Activating the Lighthouse Deck in the estimator

### DISCUSSION THREAD

**Discussion by tmartindev** [bitcraze/discussions#727] [Q&A]
Hello,

This is what I want to do :
- The drone comes into a room where a lighthouse system has been calibrated using only the flow deck or a "relative" positioning system. The lighthouse deck position and yaw estimates are not pushed to the estimator.
- When the signal is detected, the drone hovers stationary. The Lighthouse calibration is loaded on the drone and sweep data is read on the PC to calculate the position and yaw of the drone.
- The "absolute" position and yaw is then pushed to the drone estimator using send_extpos or send_extpose
- The lighthouse deck is "turned on"
- The flow deck is "turned off"
- The drone can fly in the room

## send_extpos
I have already tested sending the position with [send_extpos](https://github.com/bitcraze/crazyflie-lib-python/blob/e6a47634d9d3deca2414cbbedef25ca487ff7415/cflib/crazyflie/localization.py#L131) before enabling the Lighthouse deck. I am using the high level commander so I have to wrap send_pos and enable_lighthouse_kalman in [send_hover_setpoint](https://github.com/bitcraze/crazyflie-lib-python/blob/e6a47634d9d3deca2414cbbedef25ca487ff7415/cflib/crazyflie/commander.py#L142) or [send_velocity_world_setpoint](https://github.com/bitcraze/crazyflie-lib-python/blob/e6a47634d9d3deca2414cbbedef25ca487ff7415/cflib/crazyflie/commander.py#L111) so the drone does not fly rapidly to its "relative" position in the "absolute" frame.

```
cf.high_level_commander.go_to(0,0,0,0,5,relative=True)
time.sleep(5)
cf.commander.send_velocity_world_setpoint(0,0,0,0)
time.sleep(0.5)
cf.loc.send_pos([x,y,z])
time.sleep(0.5)
for i in range(10):
    cf.commander.send_velocity_world_setpoint(0,0,0,0)
    if i==5:
        enable_lighthouse_kalman(cf)
        time.sleep(0.5)
cf.high_level_commander.go_to(0,0,0,0,5,relative=True)
time.sleep(5)
...
```
When the lighthouse is enabled, the drone is a moves a bit in position (10-30 cm) which is OK. However, it does not keep the yawrate=0. It spins to reach the "relative" yaw in the "absolute frame.
For example, the drone takes off at "absolute" lighthouse yaw = 180° but "relative" yaw = 0°. When the deck is turned on, the drone spins fast to get to absolute yaw = 0.

I don't know how to keep the drone steady when enabling the Lighthouse deck.

## send_extpose
I have also tested to use [send_extpose](https://github.com/bitcraze/crazyflie-lib-python/blob/e6a47634d9d3deca2414cbbedef25ca487ff7415/cflib/crazyflie/localization.py#L143) (roll and pitch are the same as stateEstimate.roll/pitch) instead. However, the yaw never updates. The stateEstimate sometimes resets or diverge 10 or 20 ms right after sending the extpose (verified in the log) which can crash the drone.

Below is an example of the stateEstimate logs when the position is OK :
```
90102,-0.004514236003160477,-0.2360098659992218,1.0344878435134888,-1.4935524463653564,-0.10467555373907089,0.3237009346485138
90112,-0.005148525815457106,-0.240146204829216,1.0344983339309692,-1.555998682975769,-0.20081299543380737,0.3341519832611084
Send extpose : pos[ 1.17697843 -0.20196077  1.07835045], rot[ -1.55599868  -0.200813   175.64110553]
90122,-0.008690953254699707,-0.2446257472038269,1.0345380306243896,-1.6283153295516968,-0.293724924325943,0.3476237654685974
90132,1.2374119758605957,-0.18224215507507324,1.0377789735794067,-1.6149468421936035,-0.2339085340499878,0.33992108702659607
90142,1.2964065074920654,-0.1630028635263443,1.0389639139175415,-1.5338085889816284,-0.17251554131507874,0.30351412296295166
90152,1.3561214208602905,-0.14348852634429932,1.0394023656845093,-1.387948751449585,-0.15266545116901398,0.27354663610458374
90162,1.412629246711731,-0.1255788803100586,1.041635274887085,-0.6973548531532288,0.3767431676387787,0.2579631209373474
send_velocity_world_setpoint: (0, 0, 0, 0)
90172,1.468624472618103,-0.10719765722751617,1.0412951707839966,0.5443933010101318,1.1385855674743652,0.244424968957901
90182,1.522452712059021,-0.08975591510534286,1.043347954750061,2.0288212299346924,2.2890405654907227,0.21637804806232452
90192,1.5746557712554932,-0.07349962741136551,1.0458576679229736,3.709737539291382,3.646829843521118,0.1474967747926712
90202,1.6273123025894165,-0.05543750524520874,1.048015832901001,5.335958957672119,5.0284929275512695,0.018766596913337708
90212,1.6773579120635986,-0.03979029878973961,1.050951600074768,6.922019958496094,6.4779582023620605,-0.15255172550678253
```

When the position is not OK (estimator may have restarted).
```
90702,1.0298388004302979,-0.7923766374588013,1.2853209972381592,-12.566697120666504,-13.510848999023438,-0.9628839492797852
90712,1.01935875415802,-0.8363542556762695,1.291869878768921,-13.283157348632812,-14.543560028076172,-1.1549626588821411
Send extpose : pos[ 1.17697843 -0.20196077  1.07835045], rot[-13.28315735 -14.54356003 175.64110553]
90722,1.0089834928512573,-0.8801226615905762,1.2980308532714844,-13.974842071533203,-15.435646057128906,-1.3009238243103027
90732,8.04636329121422e-06,-5.2395789680304006e-06,-3.239991201553494e-05,-0.18224309384822845,-0.32002055644989014,0.03157404065132141
90742,-6.720100373058813e-06,-9.546231012791395e-05,1.414676308631897,-0.6335801482200623,-1.0179473161697388,0.0853566974401474
90752,-7.896982424426824e-06,-0.00016587672871537507,1.413977026939392,-0.9766750335693359,-1.5888363122940063,0.05756501853466034
90762,8.85849294718355e-06,-0.00023028990835882723,1.413065791130066,-1.1760478019714355,-2.004004955291748,0.015577242709696293
send_velocity_world_setpoint: (0, 0, 0, 0)
90772,4.242565773893148e-05,-0.00029236008413136005,1.4120469093322754,-1.2504366636276245,-2.2830827236175537,-0.004176119342446327
90782,0.00011272636766079813,-0.0003212796291336417,1.4109923839569092,-1.2584359645843506,-2.4534380435943604,2.3938857339089736e-05
90792,0.0002355048491153866,-0.00034992076689377427,1.434096097946167,-1.2558469772338867,-2.5203890800476074,0.014761623926460743
90802,0.00040411201189272106,-0.00033452920615673065,1.433477759361267,-1.2369712591171265,-2.518096685409546,0.028627783060073853
90812,0.0006225945544429123,-0.0002920096449088305,1.433140754699707,-1.2989777326583862,-2.457814931869507,0.05736372619867325
```

**Comment by krichardsson** [comment]
This is an interesting experiment, sounds fun!

My interpretation is that you have two phases that you want to transition between:
1. Relative positioning using the flow deck
2. Absolution positioning using the lighthouse

Let's start by looking at the estimator. It is important to remember that the kalman estimator will maintain an absolute position estimate (in the global reference frame) regardless of if you use the flow or lighthouse deck. The only difference is really what input data that is sent to the estimator. This means that if you take off somewhere using the flow deck, the estimator will use the flow information to track the change in position, and the estimated position will be based on the position the estimator had at take off + the delta. 
When you get into the lighthouse system, absolute position and yaw information will be available from the lighthouse system, and if this does not match the current estimate, the estimate will be "pushed" to what the lighthouse system sees fairly quickly (with the default settings). 

I see 4 ways of doing this transition in a "nice" way:
1. Try to align the coordinate system when you take off as much as possible to be close to the absolute coordinate system defined in the lighthouse system. This can be done by resetting the estimator to specific values when you take off.
2. Update the lighthouse geometry to match the current estimated pose before switching to the lighthouse system. This should include a rotation and translation of the base station poses.
3. Switch to the lighthouse system "softly" to let the estimator adapt to the lighthouse coordinate system less abruptly. This could maybe be done by using a large standard deviation initially, but I'm not sure it would work.
4. Update the estimated position in the estimator mid flight. 

4 is what I would suggest, and I think that this is what you are trying to do (or a bit of 3).

What I think you need to is to change step **The "absolute" position and yaw is then pushed to the drone estimator using send_extpos or send_extpose"**. When you use `send_extpos()` you actually input information to the estimator (like any other sensor information) and this will make the estimator to move the estimated position closer to this position, but it will also update velocities and attitude in the estimated state, which not is what you want (controllers will be confused). 
Instead of using `send_extpos()` you should reset the estimator to the new global coordinate system. You can do this by first setting the parameters
[kalman.initialX](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#kalmaninitialx)
[kalman.initialY](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#kalmaninitialy)
[kalman.initialZ](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#kalmaninitialz)
[kalman.initialYaw](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#kalmaninitialyaw)

and then set the parameter [kalman.resetEstimation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#kalmanresetestimation) to 1

It is probably a good idea to turn off the high level commander and send a velocity setpoint with zero velocity first.
